lib/streamlit/DeltaGenerator.py

```diff
@@ -132,14 +160,28 @@ def _set_widget_id(widget_type, element):
         The type of the widget as stored in proto.
     element : proto
         The proto of the element
+    user_key : str
+        Optional user-specified key to use for the widget ID.
+        If this is None, we'll generate an ID by hashing the element.
 
     """
-    widget_id = "%s" % hash(element.SerializeToString())
+    if user_key is not None:
+        widget_id = "%s-%s" % (widget_type, user_key)
+    else:
+        widget_id = "%s" % hash(element.SerializeToString())
+
+    ctx = get_report_ctx()
+    if ctx is not None:
+        added = ctx.widget_ids_this_run.add(widget_id)
+        if not added:
```

Comment:
It is not straightforward that `added is None` means that the widget id is duplicated. Maybe raise the exception inside `add(`?

Reply:
I (softly!) disagree - I have `add()` returning a bool (not a `None`) indicating whether the item was successfully added, which mirrors the API in other languages (e.g. https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.add?view=netframework-4.8).

I'm hesitant to add an Exception to indicate this case; it feels like it violates the "Never use exception for flow-control" principle. The item existing in the set already is _not_ (IMO) an exceptional case - the exception case is when the user has defined a duplicate key, but this is the concern of the DeltaGenerator function, not the set it uses to detect this.

That said, I don't disagree so strongly that I wouldn't back down! Does anyone else in @streamlit/eng have an opinion here?

Reply:
I think I come to agree with you. I missed it is returning a `bool`.